clients/unshaded/src/main/java/tachyon/client/block/TachyonBlockStore.java

```diff
@@ -89,25 +90,29 @@ public BufferedBlockInStream getInStream(long blockId) throws IOException {
     try {
       // TODO(calvin): Fix this RPC.
       BlockInfo blockInfo = masterClient.getBlockInfo(blockId);
-      // TODO(calvin): Get location via a policy.
       if (blockInfo.locations.isEmpty()) {
-        // TODO(calvin): Maybe this shouldn't be an exception.
-        throw new IOException("No block " + blockId + " is not available in Tachyon");
+        throw new IOException("Block " + blockId + " is not available in Tachyon");
       }
-      // TODO(calvin): Investigate making this a Factory method
-      NetAddress workerNetAddress = blockInfo.locations.get(0).getWorkerAddress();
-      InetSocketAddress workerAddr =
-          new InetSocketAddress(workerNetAddress.getHost(), workerNetAddress.getDataPort());
-      if (NetworkAddressUtils.getLocalHostName(ClientContext.getConf()).equals(
-          workerAddr.getHostName())) {
-        if (mContext.hasLocalWorker()) {
-          return new LocalBlockInStream(blockId, blockInfo.getLength(), workerAddr);
-        } else {
-          throw new IOException("Local read requested but there is no local worker.");
+      // TODO(calvin): Get location via a policy.
+      // Although blockInfo.locations are sorted by tier, we prefer reading from the local worker.
+      // But when there is no local worker or there are no local blocks, we prefer the first
+      // location in blockInfo.locations that is nearest to memory tier.
+      // Assuming if there is no local worker, there are no local blocks in blockInfo.locations.
+      if (mContext.hasLocalWorker()) {
```

Comment:
If the local worker has died, then `hasLocalWorker()` might still be `true` depending on when the client starts, but even in that case, I think it's fine to check the flag because if there is actually no local worker, we can spare the time of traversing and the string comparison inside the loop.